YHP_website/src/main/java/storage/UserCache.java

```diff
@@ -36,46 +38,67 @@ private UserCache(final Database database)
 	{
-		QueryRunner run = new QueryRunner(Database.getInstance().getDataSource());
-		List<User> users = null;
+		QueryRunner run = this.database.getQueryRunner();
+		List<User> foundUsers = null;
 		try
 		{
-			users = run.query(
-					"SELECT * FROM users WHERE user_name='?';",
-					new UserResultsSetHandler(), username);
+			foundUsers = run.query("SELECT * FROM users WHERE ?=?;",
+					new UserResultsSetHandler(), columnName, columnValue);
 		}
 		catch (SQLException e)
 		{
 			e.printStackTrace();
 		}
-		if (users == null || users.isEmpty())
+		if (foundUsers == null || foundUsers.isEmpty())
 		{
 			return null;
 		}
-		return users.get(0);
+		this.users.addAll(foundUsers);
+		return foundUsers.get(0);
+	}
+	
+	public void putUser(final User user) throws SQLException
+	{
+		AsyncQueryRunner run = this.database.getAsyncQueryRunner();
+		run.update("INSERT INTO users (user_id, user_name, user_pass, forename, surname, address_id) VALUES (?,?,?,?,?,?)"
+		+ "ON CONFLICT (user_id) DO UPDATE SET user_name = EXCLUDED.user_name, user_pass=EXCLUDED.user_pass, forename=EXCLUDED.forename, surname=EXCLUDED.surname, address_id=EXCLUDED.address_id;",
+				user.getUuid(), user.getUsername(), user.getDigest(), user.getForename(), user.getSurname(), user.getAddress().getId());
+		for (final String rolename : user.getRole().toDatabaseValues())
+		{
+			run.update("INSERT INTO user_roles (user_name, role_name) VALUES (?,?) ON CONFLICT (user_name, role_name) DO NOTHING", user.getUsername(), rolename);
+		}
 	}
 	
-	public boolean putUser(final User user)
+	public boolean isUsernameTaken(final String username)
 	{
-		QueryRunner run = new QueryRunner(this.database.getDataSource());
-		return false;
+		return this.getUser(username) == null;
 	}
 	
 	private class UserResultsSetHandler extends AbstractListHandler<User>
 	{
 		@Override
 		protected User handleRow(final ResultSet rs) throws SQLException
 		{
-			final User.Builder userBuilder =
-					new User.Builder(rs.getString("forename"), rs.getString("surname"));
-			final QueryRunner roleRunner = new QueryRunner(Database.getInstance().getDataSource());
+			final User.Builder userBuilder = new User.Builder(rs.getString("forename"), rs.getString("surname"));
+			final QueryRunner roleRunner = UserCache.this.database.getQueryRunner();
+			userBuilder.username(rs.getString("user_name"));
+			userBuilder.digest(new Digest(rs.getString("user_pass")));
+			userBuilder.id(UUID.fromString(rs.getString("user_id")));
 			final List<String> roles = roleRunner.query(
-					"SELECT role_name FROM user_roles WHERE user_name='?';",
+					"SELECT role_name FROM user_roles WHERE user_name=?;",
 					new ColumnListHandler<String>(), rs.getString("user_name"));
-			userBuilder.id(UUID.fromString(rs.getString("user_id")));
 			userBuilder.role(UserRole.fromDatabaseValues(roles));
 			// userBuilder.address(value); TODO add address getter.
 			return userBuilder.build();

```

Comment:
I can't immediately tell what this all does, a little explanation would be handy